src/batch/src/task/task_execution.rs

```diff
@@ -331,25 +334,33 @@ impl<C: BatchTaskContext> BatchTaskExecution<C> {
         Ok(())
     }
 
-    /// Change state and notify frontend for task status.
+    /// Change state and notify frontend for task status via streaming GRPC.
     pub async fn change_state_notify(
         &self,
         task_status: TaskStatus,
         state_tx: &mut tokio::sync::mpsc::Sender<TaskInfoResponseResult>,
+        err_str: Option<String>,
     ) -> BatchResult<()> {
         self.change_state(task_status);
-        // Notify frontend the task status.
-        state_tx
-            .send(Ok(TaskInfoResponse {
-                task_info: Some(TaskInfo {
-                    task_id: Some(TaskId::default().to_prost()),
-                    task_status: task_status.into(),
-                }),
-                // TODO: Fill the real status.
-                ..Default::default()
-            }))
-            .await
-            .map_err(|_| SenderError)
+        if let Some(err_str) = err_str {
+            state_tx
+                .send(Err(RwError::from(anyhow!(err_str)).into()))
```

Comment:
1. It's better to have another error type rather than general error.
2. If task failure is no longer used, we should remove it to make code cleaner.

Reply:
Fix 1.
For 2, Yes we can remove the TaskStatus::Failed, but the task execution may stay alive for some time even after it is claimed as Failed due to observation, so maybe this is still useful